deps/rabbit/test/quorum_queue_prop_SUITE.erl

```diff
@@ -0,0 +1,1089 @@
+%% License, v. 2.0. If a copy of the MPL was not distributed with this
+%% file, You can obtain one at https://mozilla.org/MPL/2.0/.
+%%
+%% Copyright (c) 2007-2021 VMware, Inc. or its affiliates.  All rights reserved.
+%%
+
+-module(quorum_queue_prop_SUITE).
+-compile(export_all).
+
+-define(NUM_TESTS, 500).
+
+%% Set to true to get an awful lot of debug logs.
+-if(false).
+-define(DEBUG(X,Y), logger:debug("~0p: " ++ X, [?FUNCTION_NAME|Y])).
+-else.
+-define(DEBUG(X,Y), _ = X, _ = Y, ok).
+-endif.
+
+-include_lib("amqp_client/include/amqp_client.hrl").
+-include_lib("proper/include/proper.hrl").
+
+-record(qq, {
+    amq = undefined :: amqqueue:amqqueue(),
+    name :: atom(),
+
+    %% We have one queue per way of publishing messages (such as channels).
+    %% We can only confirm the publish order on a per-channel level because
+    %% the order is non-deterministic when multiple publishers concurrently
+    %% publish.
+    q = #{} :: #{pid() | internal => queue:queue()},
+
+    %% We must account for lost acks/unconfirmed publishes after restarts.
+    restarted = false :: boolean(),
+    crashed = false :: boolean(),
+    %% We must keep some received messages around because basic.ack is not
+    %% synchronous and as a result we may end up receiving messages twice
+    %% after a queue is restarted.
+    acked = [] :: list(),
+    %% We keep track of persistent messages that were confirmed and for
+    %% which we have received the publisher confirm. This is used when
+    %% restarting the queue to only move messages into 'uncertain'
+    %% when we have not received confirms.
+    confirmed = [] :: list(),
+    %% We also keep track of persistent messages that were sent before
+    %% the channel entered confirm mode. These messages will not be
+    %% confirmed retroactively and must not be considered confirmed.
+    unconfirmed = [] :: list(),
+    %% We must also keep some published messages around because when
+    %% confirms are not used, or were not received, we are uncertain
+    %% of whether the message made it to the queue or not.
+    uncertain = [] :: list(),
+    %% We may receive any message previously received after a crash.
+    received = [] :: list(),
+
+    %% CT config necessary to open channels.
+    config = no_config :: list(),
+
+    %% Limiter pid for calling rabbit_amqqueue:basic_get.
+    limiter :: pid(),
+
+    %% Channels.
+    %%
+    %% We set consumers to 'true' when there had been consumers, even if they
+    %% were closed or cancelled, because there can be a race condition between
+    %% the cancelling and the message going back to the queue, and the test
+    %% suite getting messages via basic.get.
+    consumers = false :: boolean(),
+    channels = #{} :: #{pid() => #{consumer := none | binary(), confirms := boolean()}}
+}).
+
+%% Common Test.
+
+all() ->
+    [{group, quorum_queue_tests}].
+
+groups() ->
+    [{quorum_queue_tests, [], [
+        manual
+%        quorum_queue
```

Comment:
This currently runs the `manual` test which reproduces a bug. To run the test suite, comment `manual` and uncomment `quorum_queue`.